processor/src/advice/providers.rs

```diff
@@ -106,6 +109,63 @@ where
         Ok(())
     }
 
+    fn falcon_sign(&self, pub_key: Word, msg: Word) -> Result<Vec<Felt>, ExecutionError> {
+        let pk_sk = self
+            .map
+            .get(&pub_key.into_bytes())
+            .ok_or(ExecutionError::AdviceKeyNotFound(pub_key))?;
+
+        if pk_sk.len() != (PK_LEN + SK_LEN) {
+            return Err(ExecutionError::AdviceStackReadFailed(0));
+        }
+
+        // To generate a signature, we need the expanded key as well as the secret key
+        let pk_exp = pk_sk[..PK_LEN].to_vec();
+        let pk_exp: PublicKeyBytes = vec_felt_to_u8(&pk_exp)
+            .try_into()
+            .expect("Should not fail as we've checked the length of the combined vector");
```

Comment:
I would probably move most of the code from this function to a separate module - maybe `src/advice/dsa.rs`. Basically, the code here would just get the private key and invoke a function in that module which would return the signature.